CONTRIBUTING.md

```diff
@@ -0,0 +1,77 @@
+  - Any plagiarized work will not be merged.
+- Your work will be distributed under [MIT License](LICENSE.md) once your pull request is merged
+- Your submitted work fulfils or mostly fulfills our styles and standards
+
+__New implementation__ is welcome! For example, new solutions for a problem, different representations for a graph data structure or algorithm designs with different complexity but __identical implementation__ of an existing implementation is not allowed. Please check whether the solution is already implemented or not before submitting your pull request.
+
+__Improving comments__ and __writing proper tests__ are also highly welcome.
+
+### Contribution
+
+We appreciate any contribution, from fixing a grammar mistake in a comment to implementing complex algorithms. Please read this section if you are contributing your work.
+
+Please help us keep our issue list small by adding fixes: Add the number of the issue you solved — even if only partially — to the commit message of your pull request. GitHub will use this tag to auto-close the issue when the PR is merged.
+
+#### What is an Algorithm?
+
+An Algorithm is one or more functions (or classes) that:
+* take one or more inputs,
+* perform some internal calculations or data manipulations,
+* return one or more outputs,
+* have minimal side effects (Examples of side effects: `print()`, `plot()`, `read()`, `write()`).
+
+Algorithms should be packaged in a way that would make it easy for readers to put them into larger programs.
+
+Algorithms should:
+* have intuitive class and function names that make their purpose clear to readers
+* use Nim naming conventions and intuitive variable names with correct typing to ease comprehension
+* Prefer `Natural` type to `uint` unless wrapping is mandatory (for e.g. modular algorithms).
+* raise Nim exceptions (`ValueError`, etc.) on erroneous input values
+* add the exceptions raised to the list of side effects
+* have docstrings with clear explanations and/or URLs to source materials
+* contain doctests that test both valid and erroneous input values
+* return all calculation results instead of printing or plotting them
+
+Algorithms in this repo should not be how-to examples for existing Nim packages. Instead, they should perform internal calculations or manipulations to convert input values into different output values.  Those calculations or manipulations can use objects or functions of existing Nim modules but each algorithm in this repo should add unique value.
+
+#### Coding Style
+
+We want your work to be readable by others; therefore, we encourage you to note the following:
+
+- Please write in Nim 2.0.
+- Please focus hard on the naming of functions, objects, and variables.  Help your reader by using __descriptive names__ that can help you to remove redundant comments.
+  - Single letter variable names must be avoided at all costs. You can use `index` instead of `i` when iterating in a `for` loop.
```

Comment:
I feel this is too harsh a constraint. when you have multiple indices `i`, `j`, ... that will appear as indices (summed or transformed in some way) it might make code unreadable. I would find more appropriate a recommendation similar to the one in Python repository:
`Single letter variable names are old school so please avoid them unless their life only spans a few lines.`

Reply:
I find very vague the "only spans a few lines" and did not want to review dozens of code with imprecise, bad variable naming.
Actually too long names can make a code more dense and hard to read indeed.

I might list a few use cases:
multiple indices (matrix, sums integrals and other variables from mathematical formula given as a reference) and (c for character if it is the only character in the scope for string algorithms).

Reply:
I was very explicit about this use, and relax greatly the use of single letter variable names.
This section is a little bit long, now !